### ISSUE TITLE
[PERF] - High CPU usage causing regular desync issues on some relays.

### ISSUE THREAD

**Issue by jbgi** [input-output-hk/cardano-node#2022]
**Internal**

**Summary**
We have two relays on mainnet that expose this behavior (rel-f-1 and rel-c-1).

**System info (please complete the following information):**
- OS: NixOS 
- Version: 20.03

- Node version: 1.21.1

**Screenshots and attachments**
Profiling results:
```
Mon Oct 26 16:43 2020 Time and Allocation Profiling Report  (Final)

	   cardano-node +RTS -T -I0 -N2 -A16m -P --machine-readable -tcardano-node.stats -l -pocardano-node -N4 -A10m -qg -qb -M10G -RTS run --config /nix/store/pgzl8x226ydfk8di7f8dp97bxjk0r2fk-config-33.json --database-path /var/lib/cardano-node/db-mainnet --topology /nix/store/g9v7xn5d98n726bfkh9rjyv2w65jny23-topology.yaml --host-addr 172.31.42.6 --port 3001

	total time  =      135.69 secs   (542759 ticks @ 1000 us, 4 processors)
	total alloc = 457,466,748,272 bytes  (excludes profiling overheads)

COST CENTRE MODULE                             SRC                                                  %time %alloc  ticks     bytes

compute3    Control.Iterate.SetAlgebraInternal src/Control/Iterate/SetAlgebraInternal.hs:1155:1-33   77.3   84.7  419562 387339645840
liftSTS     Control.State.Transition.Extended  src/Control/State/Transition/Extended.hs:259:1-30     10.1    8.6  55022 39160275096


                                                                                                                                                                                                                                                                               individual      inherited
COST CENTRE                                                                                          MODULE                                                          SRC                                                                                  no.       entries  %time %alloc   %time %alloc  ticks     bytes

MAIN                                                                                                 MAIN                                                            <built-in>                                                                            2154          0    0.4    0.0   100.0  100.0   1902 102704528
   runRunCommand                                                                                     Main                                                            app/cardano-node.hs:94:1-40                                                           4513          1    0.0    0.0    99.6  100.0      0        16
    runNode                                                                                          Cardano.Node.Run                                                src/Cardano/Node/Run.hs:(97,1)-(178,37)                                               4514          1    0.0    0.0    99.6  100.0      1     18040
     withShutdownHandling                                                                            Cardano.Node.Handlers.Shutdown                                  src/Cardano/Node/Handlers/Shutdown.hs:(138,1)-(152,80)                                6085          1    0.0    0.0    98.3   98.5      0       816
      run                                                                                            Ouroboros.Consensus.Node                                        src/Ouroboros/Consensus/Node.hs:(157,1)-(326,9)                                       6086          1    0.0    0.0    98.3   98.5      0      6128
       withLockDB                                                                                    Ouroboros.Consensus.Node.DbLock                                 src/Ouroboros/Consensus/Node/DbLock.hs:(27,1)-(32,19)                                 6818          1    0.0    0.0    98.3   98.5      0       104
        withLockDB_                                                                                  Ouroboros.Consensus.Node.DbLock                                 src/Ouroboros/Consensus/Node/DbLock.hs:(57,1)-(80,53)                                 6819          1    0.0    0.0    98.3   98.5      0      2488
         createMarkerOnCleanShutdown                                                                 Ouroboros.Consensus.Node.Recovery                               src/Ouroboros/Consensus/Node/Recovery.hs:(36,1)-(38,34)                               6889          1    0.0    0.0    98.3   98.5      0       552
          withRegistry                                                                               Ouroboros.Consensus.Util.ResourceRegistry                       src/Ouroboros/Consensus/Util/ResourceRegistry.hs:589:1-54                             6891        188    0.0    0.0    98.3   98.5      4    186928
           runDataDiffusion                                                                          Ouroboros.Network.Diffusion                                     src/Ouroboros/Network/Diffusion.hs:(136,1)-(352,43)                                  18259          1    0.0    0.0    96.4   96.9      0      1560
            withIOManager                                                                            System.IOManager                                                src/System/IOManager.hs:83:1-51                                                      18261          0    0.0    0.0    96.4   96.9      0      3192
              withServer                                                                             Ouroboros.Network.NodeToNode                                    src/Ouroboros/Network/NodeToNode.hs:(434,1)-(446,34)                                 18322          1    0.0    0.0    95.2   96.6      0       656
               withServerNode'                                                                       Ouroboros.Network.Socket                                        src/Ouroboros/Network/Socket.hs:(669,1)-(693,17)                                     18323          1    0.0    0.0    95.2   96.6      0      2768
                run                                                                                  Ouroboros.Network.Server.Socket                                 src/Ouroboros/Network/Server/Socket.hs:(287,1)-(299,31)                              18338          1    0.0    0.0    95.2   96.6     36    377256
                 asyncWithUnmask                                                                     Control.Concurrent.Async                                        Control/Concurrent/Async.hs:208:1-75                                                 30974        247    0.0    0.0    95.2   96.6     16    440968
                  beginConnection                                                                    Ouroboros.Network.Socket                                        src/Ouroboros/Network/Socket.hs:(355,1)-(391,62)                                     30977          0    0.0    0.0    95.2   96.6     71  70539952
                   muxStart                                                                          Network.Mux.Compat                                              src/Network/Mux/Compat.hs:(106,1)-(161,46)                                           31106        244    0.0    0.0    95.2   96.6      6   1400704
                    runMux                                                                           Network.Mux                                                     src/Network/Mux.hs:(167,1)-(185,44)                                                  31140        245    0.0    0.0    95.2   96.6      2    400592
                     withJobPool                                                                     Control.Concurrent.JobPool                                      src/Control/Concurrent/JobPool.hs:(36,1)-(48,23)                                     31142        245    0.0    0.0    95.1   96.6     33   5698480
                      forkJob                                                                        Control.Concurrent.JobPool                                      src/Control/Concurrent/JobPool.hs:(55,1)-(73,26)                                     31144       1470    0.0    0.0    95.1   96.6      5   1832376
                       async                                                                         Control.Concurrent.Async                                        Control/Concurrent/Async.hs:194:1-35                                                 31145          0    0.0    0.0    95.1   96.6     14   4297360
                        toApplication                                                                Ouroboros.Network.Mux                                           src/Ouroboros/Network/Mux.hs:(144,1)-(151,52)                                        31194          0    0.0    0.0    92.9   96.2      1     23088
                         runHandshakeServer                                                          Ouroboros.Network.Protocol.Handshake                            src/Ouroboros/Network/Protocol/Handshake.hs:(148,1)-(165,76)                         31195          0    0.0    0.0    92.9   96.2      0         0
                          runPeerWithLimits                                                          Ouroboros.Network.Driver.Limits                                 src/Ouroboros/Network/Driver/Limits.hs:(220,1)-(223,59)                              31196        735    0.0    0.0    92.9   96.2      0     17640
                           withTimeoutSerial                                                         Network.Mux.Timeout                                             src/Network/Mux/Timeout.hs:62:1-43                                                   31197        979    0.0    0.0    92.9   96.2      0         0
                            withTimeoutSerialNative                                                  Network.Mux.Timeout                                             src/Network/Mux/Timeout.hs:67:1-54                                                   31198        979    0.0    0.0    92.9   96.2      1    117240
                             runPeerWithDriver                                                       Network.TypedProtocol.Driver                                    src/Network/TypedProtocol/Driver.hs:(106,1)-(122,24)                                 31199        735    0.0    0.0    92.9   96.2      0    101752
                              handshakeServerPeer                                                    Ouroboros.Network.Protocol.Handshake.Server                     src/Ouroboros/Network/Protocol/Handshake/Server.hs:(32,1)-(72,62)                    31200          0    0.0    0.0    92.9   96.2      0         0
                               simpleSingletonVersions                                               Ouroboros.Network.Protocol.Handshake.Version                    src/Ouroboros/Network/Protocol/Handshake/Version.hs:(178,1)-(181,63)                 31201          0    0.0    0.0    92.9   96.2      0         0
                                responder                                                            Ouroboros.Consensus.Network.NodeToNode                          src/Ouroboros/Consensus/Network/NodeToNode.hs:(621,1)-(634,13)                       31202          0    0.0    0.0    92.9   96.2      0         0
                                 mkApps                                                              Ouroboros.Consensus.Network.NodeToNode                          src/Ouroboros/Consensus/Network/NodeToNode.hs:(412,1)-(574,25)                       31203          0    0.0    0.0    92.9   96.2    141  31403560
                                  runPipelinedPeerWithDriver                                         Network.TypedProtocol.Driver                                    src/Network/TypedProtocol/Driver.hs:(156,1)-(172,27)                                 31245          0    0.0    0.0    91.1   95.2      0    340608
                                   race                                                              Control.Concurrent.Async                                        Control/Concurrent/Async.hs:(675,1)-(681,31)                                         31247        245    0.0    0.0    91.1   95.2    230  54349696
                                    txSubmissionServerPeerPipelined                                  Ouroboros.Network.Protocol.TxSubmission.Server                  src/Ouroboros/Network/Protocol/TxSubmission/Server.hs:(108,1)-(157,42)               31276          0    0.0    0.0    90.1   94.9     61  22844200
                                     mkHandlers                                                      Ouroboros.Consensus.Network.NodeToNode                          src/Ouroboros/Consensus/Network/NodeToNode.hs:(161,1)-(199,7)                        31278          0    0.0    0.0    90.0   94.9      7   1664728
                                      txSubmissionInbound                                            Ouroboros.Network.TxSubmission.Inbound                          src/Ouroboros/Network/TxSubmission/Inbound.hs:(168,1)-(427,49)                       31279          0    0.1    0.0    90.0   94.9    342  67562384
                                       getMempoolWriter                                              Ouroboros.Consensus.NodeKernel                                  src/Ouroboros/Consensus/NodeKernel.hs:(591,1)-(596,5)                                32216          0    0.0    0.0    89.9   94.8      3    470560
                                        addTxs                                                       Ouroboros.Consensus.Mempool.API                                 src/Ouroboros/Consensus/Mempool/API.hs:(212,1)-(237,26)                              32219          0    0.0    0.0    89.9   94.8      8         0
                                         tryAddTxs                                                   Ouroboros.Consensus.Mempool.API                                 src/Ouroboros/Consensus/Mempool/API.hs:120:7-15                                      32238          0    0.0    0.0    89.9   94.8      0         0
                                          initNodeKernel                                             Ouroboros.Consensus.NodeKernel                                  src/Ouroboros/Consensus/NodeKernel.hs:(142,1)-(172,7)                                32239          0    0.0    0.0    89.9   94.8      0        72
                                           openMempool                                               Ouroboros.Consensus.Mempool.Impl                                src/Ouroboros/Consensus/Mempool/Impl.hs:(76,1)-(79,26)                               32240          0    0.0    0.0    89.9   94.8    146  23694368
                                            hczipWith3                                               Data.SOP.Classes                                                src/Data/SOP/Classes.hs:341:1-21                                                     32270      14910    0.0    0.0    89.1   94.4      0         0
                                             hcliftA3                                                Data.SOP.Classes                                                src/Data/SOP/Classes.hs:319:1-68                                                     32271      14882    0.0    0.0    89.1   94.4     31  11788544
                                              lift                                                   Ouroboros.Consensus.HardFork.Combinator.State.Lift              src/Ouroboros/Consensus/HardFork/Combinator/State/Lift.hs:19:1-43                    32275      29790    0.0    0.0    89.1   94.4      1         0
                                               liftM                                                 Ouroboros.Consensus.HardFork.Combinator.State.Lift              src/Ouroboros/Consensus/HardFork/Combinator/State/Lift.hs:23:1-53                    32286      14837    0.0    0.0    89.1   94.4      6    353016
                                                fn_3                                                 Data.SOP.Classes                                                src/Data/SOP/Classes.hs:147:1-55                                                     32287          0    0.0    0.0    89.1   94.4     55   5417328
                                                 overShelleyState                                    Shelley.Spec.Ledger.API.Mempool                                 src/Shelley/Spec/Ledger/API/Mempool.hs:(121,1)-(127,7)                               32288      14703    0.0    0.0    89.1   94.4      6    354648
                                                  applyTxs                                           Shelley.Spec.Ledger.API.Mempool                                 src/Shelley/Spec/Ledger/API/Mempool.hs:(105,1)-(112,13)                              32289      14697    0.0    0.0    89.1   94.4     87  42529288
                                                   applySTS                                          Control.State.Transition.Extended                               src/Control/State/Transition/Extended.hs:(328,1)-(345,9)                             32290      14787    0.0    0.0    89.1   94.4     12   4835200
                                                    applySTSOpts                                     Control.State.Transition.Extended                               src/Control/State/Transition/Extended.hs:(316,1)-(321,15)                            32291      14880    0.0    0.0    89.0   94.4     76  25824032
                                                     foldF                                           Control.Monad.Free.Church                                       src/Control/Monad/Free/Church.hs:193:1-41                                            32292      93463    0.0    0.0    89.0   94.4     19   5568232
                                                      judgmentContext                                Control.State.Transition.Extended                               src/Control/State/Transition/Extended.hs:263:1-36                                    32293          0    0.0    0.0    89.0   94.4     44  13577536
                                                       trans                                         Control.State.Transition.Extended                               src/Control/State/Transition/Extended.hs:253:1-36                                    32297      79199    0.1    0.1    89.0   94.4    609 256001320
                                                        ?!                                           Control.State.Transition.Extended                               src/Control/State/Transition/Extended.hs:238:1-89                                    32318     178403    0.0    0.0    77.7   84.8     20   5576608
                                                         liftF                                       Control.Monad.Free.Class                                        src/Control/Monad/Free/Class.hs:154:1-26                                             32319     178503    0.0    0.0    77.7   84.8    120  53498544
                                                          liftSTS                                    Control.State.Transition.Extended                               src/Control/State/Transition/Extended.hs:259:1-30                                    32425          0    0.2    0.0    77.7   84.8    875 178519704
                                                           eval                                      Control.Iterate.SetAlgebraInternal                              src/Control/Iterate/SetAlgebraInternal.hs:1278:1-29                                  32442      28195    0.0    0.0    77.3   84.7      3         0
                                                            compute                                  Control.Iterate.SetAlgebraInternal                              src/Control/Iterate/SetAlgebraInternal.hs:(1018,1)-(1150,46)                         32443      28091    0.0    0.0    77.3   84.7     37    526480
                                                             compute3                                Control.Iterate.SetAlgebraInternal                              src/Control/Iterate/SetAlgebraInternal.hs:1155:1-33                                  32948      13162   77.3   84.7    77.3   84.7  419562 387339456720
```

Attachments:
[rel-f-1.zip](https://github.com/input-output-hk/cardano-node/files/5440663/rel-f-1.zip)

**Additional context**
The profiling was done using this branch of `cardano-ledger-spec`, in order to identify the problematic case of the `compute` function:

https://github.com/input-output-hk/cardano-ledger-specs/blob/jbgi/desync-investiguation/semantics/executable-spec/src/Control/Iterate/SetAlgebraInternal.hs#L1020

which correspond to this line in latest master:
https://github.com/input-output-hk/cardano-ledger-specs/blob/d371ad2ebf5d1ddff93776fac2dfd1045aa6b06c/semantics/executable-spec/src/Control/Iterate/SetAlgebra.hs#L1086

**Comment by Jimbo4350** [comment]
@jbgi is this still relevant?

**Comment by jbgi** [comment]
was fixed by c54b3b862f9070af71ae503fc53eddf3e9538058
